### PR TITLE
Redmine # 7769: Add support for Azure DNS to DynamicDNS

### PR DESCRIPTION
**Pull request by martgras**
Feature#7769

- [x] Redmine Issue: https://redmine.pfsense.org/issues/7769
- [x] Ready for review

Adds  Azure DNS support to Dynamic DNS 
Only ipv4 tested. I'm unable to test ipv6

### CODE REVIEW

File: src/etc/inc/dyndns.class [hunk 1]
```diff
@@ -995,6 +999,76 @@
...
+				case 'azurev6':
+				case 'azure':
+					$hostname = "{$this->_dnsHost}";
+					$resourceid = trim($this->_dnsZoneID);
+					$app_id = $this->_dnsUser;
+					$client_secret = $this->_dnsPass;
+					$newip = $this->_dnsIP;
+					$newttl = $this->_dnsTTL;
+					// correct missing / in resourceid
+					if ($resourceid[0] != '/')
```
Comment (jim-p) [inline src/etc/inc/dyndns.class]: You are checking the first character of the string, and if it isn't a `/` then you add `/` to the end of the string. Shouldn't that be consistent? (Test the end and add at the end, or test the beginning and add at the beginning).

There are probably easier ways to accomplish this depending on what you're after. For example, if it always must end with exactly one slash, you can do something like this: `rtrim($resourceid, '/') . '/'`. If it must always start and end with a slash, then `'/' . trim($resourceid, '/') . '/'`.

File: src/etc/inc/dyndns.class [hunk 2]
```diff
@@ -995,6 +999,76 @@
...
+					{
+						$resourceid = $resourceid."/";
+					}
+					
+					preg_match('/\\/subscriptions\\/(?<sid>[^\\/]*)/',$resourceid,$result);
+					$subscriptionid=isset($result['sid']) ? $result['sid'] : '' ;
+					if ($subscriptionid == '')
```
Comment (jim-p) [inline src/etc/inc/dyndns.class]: Please use K&R, BSD KNF variant style ( https://doc.pfsense.org/index.php/Developer_Style_Guide#Indent_style )

File: src/etc/inc/dyndns.class [hunk 3]
```diff
@@ -995,6 +999,76 @@
...
+					}
+					curl_setopt($ch, CURLOPT_URL, "https://management.azure.com/subscriptions/".$subscriptionid."?api-version=2016-09-01");
+					curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+					curl_setopt($ch, CURLOPT_HEADER, 1);
+					curl_setopt($ch, CURLOPT_NOBODY, 1);
+					$output = curl_exec($ch);
+					$pattern='/Bearer authorization_uri="https:\\/\\/login.windows.net\\/(?<tid>[^"]*)/i';
+					preg_match($pattern,$output,$result);
+					$tenantid= (isset($result['tid']) ? $result['tid'] : '');
+					if ($tenantid == '')
```
Comment (jim-p) [inline src/etc/inc/dyndns.class]: Please use K&R, BSD KNF variant style ( https://doc.pfsense.org/index.php/Developer_Style_Guide#Indent_style )

File: src/etc/inc/dyndns.class [hunk 4]
```diff
@@ -995,6 +999,76 @@
...
+					curl_setopt($ch, CURLOPT_URL,"https://login.microsoftonline.com/".$tenantid."/oauth2/token");
+					curl_setopt($ch, CURLOPT_POST, 1);
+					$body = "resource=".urlencode("https://management.core.windows.net/")."&grant_type=client_credentials&client_id=".$app_id."&client_secret=".urlencode($client_secret);
+					curl_setopt($ch, CURLOPT_POSTFIELDS,$body);
+					curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+					$server_output = curl_exec ($ch);
+					$httpcode= curl_getinfo($ch, CURLINFO_HTTP_CODE);
+					preg_match("/\"access_token\":\"([^\"]*)\"/", $server_output, $token);
+					$bearertoken=$token[1];
+					if ($bearertoken == '')
```
Comment (jim-p) [inline src/etc/inc/dyndns.class]: Please use K&R, BSD KNF variant style ( https://doc.pfsense.org/index.php/Developer_Style_Guide#Indent_style )

File: src/etc/inc/dyndns.class [hunk 5]
```diff
@@ -995,6 +999,76 @@
...
+					{
+						return false;
+						log_error("no valid bearer token");
+					}
+					
+					if ($this->_useIPv6)
```
Comment (jim-p) [inline src/etc/inc/dyndns.class]: Please use K&R, BSD KNF variant style ( https://doc.pfsense.org/index.php/Developer_Style_Guide#Indent_style )

File: src/etc/inc/dyndns.class (hunk 1 above)
Comment (martgras) [reply src/etc/inc/dyndns.class]: thx you are right fixed along with the other issues

**Comment by sbeaver-netgate** [comment]
Please rebase/resubmit

**Comment by martgras** [comment]
closing this PR and resubmit as #3913 3926

**Pull request closed without merging**